File: tmdb-core/src/main/java/dev/reuss/tmdb/common/watchprovider/WatchProviderAvailability.java
package dev.reuss.tmdb.common.watchprovider;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import dev.reuss.tmdb.common.TmdbModel;
import dev.reuss.tmdb.domain.watchproviders.model.WatchProvider;
import dev.reuss.tmdb.util.TmdbCollections;

import java.util.List;

/**
 * Watch provider availability for a specific region.
 *
 * @param link     TMDB watch provider link
 * @param flatrate flatrate streaming providers
 * @param rent     rental providers
 * @param buy      purchase providers
 * @param ads      ad-supported providers
 * @param free     free providers
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public record WatchProviderAvailability(
        String link,
        List<WatchProvider> flatrate,
        List<WatchProvider> rent,
        List<WatchProvider> buy,
        List<WatchProvider> ads,
        List<WatchProvider> free
) implements TmdbModel {

    public WatchProviderAvailability {
        flatrate = TmdbCollections.list(flatrate);
        rent = TmdbCollections.list(rent);
        buy = TmdbCollections.list(buy);
        ads = TmdbCollections.list(ads);
        free = TmdbCollections.list(free);
    }
}

File: tmdb-core/src/main/kotlin/dev/reuss/tmdb/common/watchprovider/WatchProviderAvailability.kt
package dev.reuss.tmdb.common.watchprovider

import com.fasterxml.jackson.annotation.JsonProperty
import dev.reuss.tmdb.common.TmdbModel
import dev.reuss.tmdb.domain.watchproviders.model.WatchProvider

/**
 * Watch provider availability for a specific region.
 *
 * @property link     TMDB watch provider link
 * @property flatrate flatrate streaming providers
 * @property rent     rental providers
 * @property buy      purchase providers
 * @property ads      ad-supported providers
 * @property free     free providers
 */
@JvmRecord
data class WatchProviderAvailability(
    @all:JsonProperty("link")
    val link: String?,

    @all:JsonProperty("flatrate")
    val flatrate: List<WatchProvider> = emptyList(),

    @all:JsonProperty("rent")
    val rent: List<WatchProvider> = emptyList(),

    @all:JsonProperty("buy")
    val buy: List<WatchProvider> = emptyList(),

    @all:JsonProperty("ads")
    val ads: List<WatchProvider> = emptyList(),

    @all:JsonProperty("free")
    val free: List<WatchProvider> = emptyList()
) : TmdbModel
